File: apps/engine/quantara_engine/models/trading.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any, Optional

from sqlalchemy import DateTime, Integer, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from quantara_engine.models.base import Base
from quantara_engine.models.enums import (
    DecisionType,
    Direction,
    EntryType,
    ExitReason,
    FillSide,
    OrderIntentStatus,
    OrderStatus,
    OrderType,
    PortfolioMode,
    PositionStatus,
    SignalAction,
)
from quantara_engine.models.types import pg_enum


class Signal(Base):
    __tablename__ = "signals"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    strategy_instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    strategy_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    instrument_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    candle_timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    action: Mapped[SignalAction] = mapped_column(
        pg_enum(SignalAction, "signal_action"), nullable=False
    )
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    confidence: Mapped[Optional[Decimal]] = mapped_column(Numeric(5, 4))
    suggested_sl: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    suggested_tp: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    metadata_: Mapped[Optional[dict[str, Any]]] = mapped_column("metadata", JSONB)
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class Decision(Base):
    __tablename__ = "decisions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    strategy_instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    signal_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    instrument_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    candle_timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    decision_type: Mapped[DecisionType] = mapped_column(
        pg_enum(DecisionType, "decision_type"), nullable=False
    )
    message: Mapped[str] = mapped_column(Text, nullable=False)
    metadata_: Mapped[Optional[dict[str, Any]]] = mapped_column("metadata", JSONB)
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class OrderIntent(Base):
    __tablename__ = "order_intents"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    signal_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    strategy_instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    portfolio_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    direction: Mapped[Direction] = mapped_column(
        pg_enum(Direction, "direction"), nullable=False
    )
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    entry_type: Mapped[EntryType] = mapped_column(
        pg_enum(EntryType, "entry_type"), nullable=False
    )
    limit_price: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    stop_loss: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    take_profit: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    target_risk_amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    actual_risk_amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    signal_candle_timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    execution_candle_timestamp: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True)
    )
    risk_profile_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    status: Mapped[OrderIntentStatus] = mapped_column(
        pg_enum(OrderIntentStatus, "order_intent_status"),
        nullable=False,
        default=OrderIntentStatus.PENDING_EXECUTION,
    )
    rejection_reason: Mapped[Optional[str]] = mapped_column(Text)
    idempotency_key: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    intent_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    portfolio_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    instrument_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    direction: Mapped[Direction] = mapped_column(
        pg_enum(Direction, "direction"), nullable=False
    )
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    order_type: Mapped[OrderType] = mapped_column(
        pg_enum(OrderType, "order_type"), nullable=False
    )
    status: Mapped[OrderStatus] = mapped_column(
        pg_enum(OrderStatus, "order_status"),
        nullable=False,
        default=OrderStatus.PENDING,
    )
    broker_order_id: Mapped[Optional[str]] = mapped_column(String(100))
    submitted_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class Position(Base):
    __tablename__ = "positions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    portfolio_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    strategy_instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    instrument_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    direction: Mapped[Direction] = mapped_column(
        pg_enum(Direction, "direction"), nullable=False
    )
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    entry_price: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    current_price: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    stop_loss: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    take_profit: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    unrealized_pnl: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    status: Mapped[PositionStatus] = mapped_column(
        pg_enum(PositionStatus, "position_status"),
        nullable=False,
        default=PositionStatus.OPEN,
    )
    opened_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    closed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class Fill(Base):
    __tablename__ = "fills"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    order_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    position_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    fill_price: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    fill_quantity: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    fees: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False, default=Decimal("0"))
    slippage: Mapped[Decimal] = mapped_column(
        Numeric(18, 8), nullable=False, default=Decimal("0")
    )
    spread_cost: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0")
    )
    base_price: Mapped[Optional[Decimal]] = mapped_column(Numeric(18, 8))
    side: Mapped[FillSide] = mapped_column(
        pg_enum(FillSide, "fill_side"), nullable=False
    )
    filled_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class Trade(Base):
    __tablename__ = "trades"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    position_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    portfolio_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    strategy_instance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    strategy_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False
    )
    instrument_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    direction: Mapped[Direction] = mapped_column(
        pg_enum(Direction, "direction"), nullable=False
    )
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    entry_price: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    exit_price: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    gross_pnl: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    realized_pnl: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    fees_total: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    slippage_total: Mapped[Decimal] = mapped_column(Numeric(18, 8), nullable=False)
    spread_total: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    target_risk_amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    actual_risk_amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    exit_reason: Mapped[ExitReason] = mapped_column(
        pg_enum(ExitReason, "exit_reason"), nullable=False
    )
    duration_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    opened_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    closed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    mode: Mapped[PortfolioMode] = mapped_column(
        pg_enum(PortfolioMode, "portfolio_mode"), nullable=False
    )
    backtest_run_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: packages/db/schema/trading.ts
import {
  index,
  integer,
  jsonb,
  numeric,
  pgTable,
  text,
  timestamp,
  uuid,
  varchar,
} from "drizzle-orm/pg-core";
import {
  decisionTypeEnum,
  directionEnum,
  entryTypeEnum,
  exitReasonEnum,
  fillSideEnum,
  orderIntentStatusEnum,
  orderStatusEnum,
  orderTypeEnum,
  portfolioModeEnum,
  positionStatusEnum,
  signalActionEnum,
} from "./enums";
import { instruments } from "./instruments";
import { backtestRuns, portfolios, riskProfiles, strategyInstances } from "./portfolio";
import { strategyVersions } from "./strategies";

export const signals = pgTable(
  "signals",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    strategyInstanceId: uuid("strategy_instance_id")
      .notNull()
      .references(() => strategyInstances.id),
    strategyVersionId: uuid("strategy_version_id")
      .notNull()
      .references(() => strategyVersions.id),
    instrumentId: uuid("instrument_id")
      .notNull()
      .references(() => instruments.id),
    candleTimestamp: timestamp("candle_timestamp", {
      withTimezone: true,
      mode: "string",
    }).notNull(),
    action: signalActionEnum("action").notNull(),
    reason: text("reason").notNull(),
    confidence: numeric("confidence", { precision: 5, scale: 4 }),
    suggestedSl: numeric("suggested_sl", { precision: 18, scale: 8 }),
    suggestedTp: numeric("suggested_tp", { precision: 18, scale: 8 }),
    metadata: jsonb("metadata"),
    mode: portfolioModeEnum("mode").notNull(),
    backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
    createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("signals_strategy_instance_candle_timestamp_idx").on(
      table.strategyInstanceId,
      table.candleTimestamp,
    ),
  ],
);

export const decisions = pgTable(
  "decisions",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    strategyInstanceId: uuid("strategy_instance_id")
      .notNull()
      .references(() => strategyInstances.id),
    signalId: uuid("signal_id").references(() => signals.id),
    instrumentId: uuid("instrument_id")
      .notNull()
      .references(() => instruments.id),
    candleTimestamp: timestamp("candle_timestamp", {
      withTimezone: true,
      mode: "string",
    }).notNull(),
    decisionType: decisionTypeEnum("decision_type").notNull(),
    message: text("message").notNull(),
    metadata: jsonb("metadata"),
    mode: portfolioModeEnum("mode").notNull(),
    backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
    createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("decisions_strategy_instance_created_at_idx").on(
      table.strategyInstanceId,
      table.createdAt,
    ),
    index("decisions_created_at_idx").on(table.createdAt),
  ],
);

export const orderIntents = pgTable("order_intents", {
  id: uuid("id").primaryKey().defaultRandom(),
  signalId: uuid("signal_id")
    .notNull()
    .references(() => signals.id),
  strategyInstanceId: uuid("strategy_instance_id")
    .notNull()
    .references(() => strategyInstances.id),
  portfolioId: uuid("portfolio_id")
    .notNull()
    .references(() => portfolios.id),
  direction: directionEnum("direction").notNull(),
  quantity: numeric("quantity", { precision: 18, scale: 8 }).notNull(),
  entryType: entryTypeEnum("entry_type").notNull(),
  limitPrice: numeric("limit_price", { precision: 18, scale: 8 }),
  stopLoss: numeric("stop_loss", { precision: 18, scale: 8 }).notNull(),
  takeProfit: numeric("take_profit", { precision: 18, scale: 8 }),
  targetRiskAmount: numeric("target_risk_amount", {
    precision: 18,
    scale: 2,
  }).notNull(),
  actualRiskAmount: numeric("actual_risk_amount", {
    precision: 18,
    scale: 2,
  }).notNull(),
  signalCandleTimestamp: timestamp("signal_candle_timestamp", {
    withTimezone: true,
    mode: "string",
  }).notNull(),
  executionCandleTimestamp: timestamp("execution_candle_timestamp", {
    withTimezone: true,
    mode: "string",
  }),
  riskProfileId: uuid("risk_profile_id")
    .notNull()
    .references(() => riskProfiles.id),
  status: orderIntentStatusEnum("status").notNull().default("pending_execution"),
  rejectionReason: text("rejection_reason"),
  idempotencyKey: varchar("idempotency_key", { length: 100 }).notNull().unique(),
  mode: portfolioModeEnum("mode").notNull(),
  backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
  createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
    .notNull()
    .defaultNow(),
});

export const orders = pgTable("orders", {
  id: uuid("id").primaryKey().defaultRandom(),
  intentId: uuid("intent_id")
    .notNull()
    .references(() => orderIntents.id),
  portfolioId: uuid("portfolio_id")
    .notNull()
    .references(() => portfolios.id),
  instrumentId: uuid("instrument_id")
    .notNull()
    .references(() => instruments.id),
  direction: directionEnum("direction").notNull(),
  quantity: numeric("quantity", { precision: 18, scale: 8 }).notNull(),
  orderType: orderTypeEnum("order_type").notNull(),
  status: orderStatusEnum("status").notNull().default("pending"),
  brokerOrderId: varchar("broker_order_id", { length: 100 }),
  submittedAt: timestamp("submitted_at", { withTimezone: true, mode: "string" }),
  mode: portfolioModeEnum("mode").notNull(),
  backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
  createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
    .notNull()
    .defaultNow(),
});

export const positions = pgTable(
  "positions",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    portfolioId: uuid("portfolio_id")
      .notNull()
      .references(() => portfolios.id),
    strategyInstanceId: uuid("strategy_instance_id")
      .notNull()
      .references(() => strategyInstances.id),
    instrumentId: uuid("instrument_id")
      .notNull()
      .references(() => instruments.id),
    direction: directionEnum("direction").notNull(),
    quantity: numeric("quantity", { precision: 18, scale: 8 }).notNull(),
    entryPrice: numeric("entry_price", { precision: 18, scale: 8 }).notNull(),
    currentPrice: numeric("current_price", { precision: 18, scale: 8 }).notNull(),
    stopLoss: numeric("stop_loss", { precision: 18, scale: 8 }).notNull(),
    takeProfit: numeric("take_profit", { precision: 18, scale: 8 }),
    unrealizedPnl: numeric("unrealized_pnl", { precision: 18, scale: 2 }).notNull(),
    status: positionStatusEnum("status").notNull().default("open"),
    openedAt: timestamp("opened_at", { withTimezone: true, mode: "string" }).notNull(),
    closedAt: timestamp("closed_at", { withTimezone: true, mode: "string" }),
    mode: portfolioModeEnum("mode").notNull(),
    backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
    createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true, mode: "string" })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("positions_portfolio_status_idx").on(table.portfolioId, table.status),
  ],
);

export const fills = pgTable("fills", {
  id: uuid("id").primaryKey().defaultRandom(),
  orderId: uuid("order_id")
    .notNull()
    .references(() => orders.id),
  positionId: uuid("position_id").references(() => positions.id),
  fillPrice: numeric("fill_price", { precision: 18, scale: 8 }).notNull(),
  fillQuantity: numeric("fill_quantity", { precision: 18, scale: 8 }).notNull(),
  fees: numeric("fees", { precision: 18, scale: 4 }).notNull().default("0"),
  slippage: numeric("slippage", { precision: 18, scale: 8 }).notNull().default("0"),
  spreadCost: numeric("spread_cost", { precision: 18, scale: 4 })
    .notNull()
    .default("0"),
  basePrice: numeric("base_price", { precision: 18, scale: 8 }),
  side: fillSideEnum("side").notNull(),
  filledAt: timestamp("filled_at", { withTimezone: true, mode: "string" }).notNull(),
  mode: portfolioModeEnum("mode").notNull(),
  backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
  createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
    .notNull()
    .defaultNow(),
});

export const trades = pgTable(
  "trades",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    positionId: uuid("position_id")
      .notNull()
      .references(() => positions.id),
    portfolioId: uuid("portfolio_id")
      .notNull()
      .references(() => portfolios.id),
    strategyInstanceId: uuid("strategy_instance_id")
      .notNull()
      .references(() => strategyInstances.id),
    strategyVersionId: uuid("strategy_version_id")
      .notNull()
      .references(() => strategyVersions.id),
    instrumentId: uuid("instrument_id")
      .notNull()
      .references(() => instruments.id),
    direction: directionEnum("direction").notNull(),
    quantity: numeric("quantity", { precision: 18, scale: 8 }).notNull(),
    entryPrice: numeric("entry_price", { precision: 18, scale: 8 }).notNull(),
    exitPrice: numeric("exit_price", { precision: 18, scale: 8 }).notNull(),
    grossPnl: numeric("gross_pnl", { precision: 18, scale: 2 }).notNull(),
    realizedPnl: numeric("realized_pnl", { precision: 18, scale: 2 }).notNull(),
    feesTotal: numeric("fees_total", { precision: 18, scale: 4 }).notNull(),
    slippageTotal: numeric("slippage_total", { precision: 18, scale: 8 }).notNull(),
    spreadTotal: numeric("spread_total", { precision: 18, scale: 4 }).notNull(),
    targetRiskAmount: numeric("target_risk_amount", {
      precision: 18,
      scale: 2,
    }).notNull(),
    actualRiskAmount: numeric("actual_risk_amount", {
      precision: 18,
      scale: 2,
    }).notNull(),
    exitReason: exitReasonEnum("exit_reason").notNull(),
    durationSeconds: integer("duration_seconds").notNull(),
    openedAt: timestamp("opened_at", { withTimezone: true, mode: "string" }).notNull(),
    closedAt: timestamp("closed_at", { withTimezone: true, mode: "string" }).notNull(),
    mode: portfolioModeEnum("mode").notNull(),
    backtestRunId: uuid("backtest_run_id").references(() => backtestRuns.id),
    createdAt: timestamp("created_at", { withTimezone: true, mode: "string" })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("trades_strategy_version_closed_at_idx").on(
      table.strategyVersionId,
      table.closedAt,
    ),
  ],
);
